### ISSUE TITLE
System.Runtime.TieredCompilation configuration knob does not work to turn off tiered compilation

### ISSUE THREAD

**Issue by jakobbotsch** [dotnet/coreclr#19679]
I can't seem to get this configuration knob working. I am using my issue reported in #19599 as a test for whether tiered compilation is enabled or not. The following PowerShell session illustrates my confusion:
```
D:\temp\test> cat .\test.csproj
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.1</TargetFramework>
    <TieredCompilation>false</TieredCompilation>
  </PropertyGroup>

</Project>

D:\temp\test> cat .\Program.cs
struct S0
{
    public byte F0;
}

public class Program
{
    static S0 s_2;
    static long s_5;
    public static void Main()
    {
        s_2.F0 = 128;
        M7(s_2);
        System.Console.WriteLine(s_5);
    }

    static void M7(S0 arg0)
    {
        s_5 = System.Threading.Volatile.Read(ref arg0.F0);
    }
}

D:\temp\test> dotnet publish -c Debug; dotnet publish -c Release
D:\temp\test> cat .\bin\Release\netcoreapp2.1\publish\test.runtimeconfig.json
{
  "runtimeOptions": {
    "tfm": "netcoreapp2.1",
    "framework": {
      "name": "Microsoft.NETCore.App",
      "version": "2.1.0"
    },
    "configProperties": {
      "System.Runtime.TieredCompilation": false
    }
  }
}

D:\temp\test> CoreRun.exe .\bin\Debug\netcoreapp2.1\publish\test.dll
128
D:\temp\test> CoreRun.exe .\bin\Release\netcoreapp2.1\publish\test.dll
128
D:\temp\test> $env:COMPlus_TieredCompilation="0"
D:\temp\test> CoreRun.exe .\bin\Debug\netcoreapp2.1\publish\test.dll
128
D:\temp\test> CoreRun.exe .\bin\Release\netcoreapp2.1\publish\test.dll
4294967168
```
The CoreRun used here is compiled from current master (5d2ae2f), and the dotnet CLI is 2.1.400.

Is this a bug, or is there some other way to turn off tiered compilation for an app without requiring the COMPlus_TieredCompilation environment variable to be set correctly before running the app?

**Comment by jakobbotsch** [comment]
Ok, after some more tinkering I realize that this is because I am using the corerun host. Is the only way to turn off tiered compilation when using corerun with the environment variable? Is there a way to turn off tiered compilation programmatically for individual assemblies?

**Comment by kouvel** [comment]
Yes for CoreRun the only way to configure runtime options is through environment variables. You can try installing a recent preview SDK for master branch 3.0.x runtime from here: https://github.com/dotnet/core-sdk/blob/master/README.md#installers-and-binaries. Set the TargetFramework to netcoreapp3.0 to use the 3.0 preview runtime instead of 2.1.

However, it looks like the latest preview SDK downloadable from that link has coreclr.dll as of a master commit from Aug 16 and tiered compilation was enabled by default in master on Aug 17, so as of that build tiered compilation is still disabled by default.

If you want to use the latest coreclr build, you can try the following:
- Under the dotnet install folder, back up shared\Microsoft.NETCore.App\3.0.0-preview1-26827-01 (or whichever 3.0.* runtime is installed) to somewhere else
- Copy and overwrite <coreclrRepoRoot>\bin\Product\Windows_NT.x64.Release\* into the folder above

With that it should use your built coreclr bits on top of the same other versions of the SDK. The builds should be close enough that they would work together, but sometimes with changes that break compatibility between repos it may not work.

**Comment by kouvel** [comment]
To configure runtime options though the .json file (or .csproj) you'd have to use dotnet (it passes config info to coreclr). CoreRun from coreclr repo does not have any knowledge of the above two files, those are managed by the CLI tools.

**Comment by jakobbotsch** [comment]
Thank you for the information. Using the environment variables is fine for my own use, but ideally for other people normal hosts would 'just' work with my app. However I would like to add an error if tiering is turned on when my app is run (I spent a couple of hour yesterday debugging a Linux machine until I realized that tiering was enabled there). Is there a good way to check whether tiering is enabled? In other words:
* In my own environments where I run with corerun, the environment variable needs to be set
* In other environments with smarter hosts, tiering can be disabled through the runtime config

Is there a better way than checking the name of the host and then either the environment variable or runtime config?
